File: backend/internal/dtos/superadmin.go
package dtos

import "time"

// ============================================================================
// Superadmin DTOs — wire format for `/admin/...` endpoints (PDF originally
// said `/superadmin/`, renamed to match frontend `(user)/admin/` convention).
//
// Conventions:
//   - status_pembayaran filter is 'Lunas' (codebase convention; PDF said 'paid').
//   - status_pesanan='aktif' in PDF means status='processing' here (narrow).
//   - Field names camelCase to match frontend types.ts.
// ============================================================================

// --- Dashboard --------------------------------------------------------------

type SuperStatistikUmumResponse struct {
	RevenueToday     float64 `json:"revenueToday"`
	RevenueThisMonth float64 `json:"revenueThisMonth"`
	OrderToday       int     `json:"orderToday"`
	AverageRating    float64 `json:"averageRating"`
}

type SuperStatistikLayananResponse struct {
	IDLayanan    int     `json:"idLayanan"`
	NamaLayanan  string  `json:"namaLayanan"`
	TotalPesanan int     `json:"totalPesanan"`
	TotalRevenue float64 `json:"totalRevenue"`
}

type SuperLiveOrderResponse struct {
	IDPesanan       int       `json:"id"`
	NamaPelanggan   string    `json:"customer"`
	NamaCabang      string    `json:"branch"`
	Status          string    `json:"status"`
	EstimasiSelesai time.Time `json:"deadline"`
	TotalHarga      float64   `json:"total"`
}

type SuperTopCabangResponse struct {
	IDCabang     int     `json:"idCabang"`
	NamaCabang   string  `json:"namaCabang"`
	TotalOrder   int     `json:"totalOrder"`
	TotalRevenue float64 `json:"totalRevenue"`
}

type SuperPaymentMixResponse struct {
	Metode       string  `json:"metode"`
	TotalEntries int     `json:"totalEntries"`
	Persentase   float64 `json:"persentase"`
}

type SuperLiveOrdersPerCabangResponse struct {
	IDCabang        int    `json:"idCabang"`
	NamaCabang      string `json:"namaCabang"`
	TotalLiveOrders int    `json:"totalLiveOrders"`
}

// --- Orders -----------------------------------------------------------------

type SuperPesananResponse struct {
	IDPesanan              int       `json:"id"`
	JumlahItemPesanan      int       `json:"jumlahItem"`
	StatusPesanan          string    `json:"status"`
	CatatanPesanan         string    `json:"catatan"`
	EstimasiSelesaiPesanan time.Time `json:"estimasiSelesai"`
	TotalHargaPesanan      float64   `json:"totalHarga"`
	PelangganID            int       `json:"pelangganId"`
	VoucherID              *int      `json:"voucherId"`
	PegawaiID              int       `json:"pegawaiId"`
	NamaPelanggan          string    `json:"namaPelanggan"`
	NoTelpPelanggan        string    `json:"noTelpPelanggan"`
	NamaPegawai            string    `json:"namaPegawai"`
	NamaCabang             string    `json:"namaCabang"`
	IDCabang               int       `json:"idCabang"`
}

type SuperOrdersStatsResponse struct {
	Total      int `json:"total"`
	Pickup     int `json:"pickup"`
	Processing int `json:"processing"`
	Delivery   int `json:"delivery"`
	Completed  int `json:"completed"`
}

// --- Voucher ----------------------------------------------------------------

type SuperVoucherResponse struct {
	ID            int       `json:"id"`
	Kode          string    `json:"kode"`
	TipeDiskon    string    `json:"tipeDiskon"`
	NilaiDiskon   float64   `json:"nilaiDiskon"`
	MinPembelian  float64   `json:"minPembelian"`
	BerlakuHingga time.Time `json:"berlakuHingga"`
	Kuota         int       `json:"kuota"`
	Terpakai      int       `json:"terpakai"`
}

type SuperVoucherStatResponse struct {
	ActiveVoucherWeek         int     `json:"activeVoucherWeek"`
	TotalCustomerSave         float64 `json:"totalCustomerSave"`
	WaktuKadaluarsaTerdekat   string  `json:"waktuKadaluarsaTerdekat"` // ISO 8601 duration-ish string
	WaktuKadaluarsaTerdekatHrs float64 `json:"waktuKadaluarsaTerdekatHours"`
}

type CreateVoucherRequest struct {
	Kode          string    `json:"kode" binding:"required"`
	TipeDiskon    string    `json:"tipeDiskon" binding:"required"`
	NilaiDiskon   float64   `json:"nilaiDiskon" binding:"required"`
	MinPembelian  float64   `json:"minPembelian"`
	BerlakuHingga time.Time `json:"berlakuHingga" binding:"required"`
	Kuota         int       `json:"kuota" binding:"required"`
}

// --- Staffs -----------------------------------------------------------------

type SuperPegawaiResponse struct {
	ID         int    `json:"id"`
	Nama       string `json:"nama"`
	Email      string `json:"email"`
	NoTelp     string `json:"noTelp"`
	Alamat     string `json:"alamat"`
	CabangID   int    `json:"cabangId"`
	CabangNama string `json:"cabangNama"`
}

type CreatePegawaiRequest struct {
	Nama     string `json:"nama" binding:"required"`
	Email    string `json:"email" binding:"required,email"`
	NoTelp   string `json:"noTelp" binding:"required"`
	Alamat   string `json:"alamat" binding:"required"`
	CabangID int    `json:"cabangId" binding:"required"`
}

type UpdatePegawaiRequest struct {
	Nama     *string `json:"nama"`
	Email    *string `json:"email" binding:"omitempty,email"`
	NoTelp   *string `json:"noTelp"`
	Alamat   *string `json:"alamat"`
	CabangID *int    `json:"cabangId"`
}

// --- Couriers ---------------------------------------------------------------

type SuperKurirResponse struct {
	ID               int    `json:"id"`
	Nama             string `json:"nama"`
	NoTelp           string `json:"noTelp"`
	NoPlat           string `json:"noPlat"`
	TipeKendaraanID  int    `json:"tipeKendaraanId"`
	JenisKendaraan   string `json:"jenisKendaraan"`
}

type CreateKurirRequest struct {
	Nama            string `json:"nama" binding:"required"`
	NoTelp          string `json:"noTelp" binding:"required"`
	NoPlat          string `json:"noPlat" binding:"required"`
	TipeKendaraanID int    `json:"tipeKendaraanId" binding:"required"`
}

type UpdateKurirRequest struct {
	Nama            *string `json:"nama"`
	NoTelp          *string `json:"noTelp"`
	NoPlat          *string `json:"noPlat"`
	TipeKendaraanID *int    `json:"tipeKendaraanId"`
}

type SuperTipeKendaraanResponse struct {
	ID              int    `json:"id"`
	JenisKendaraan  string `json:"jenisKendaraan"`
}

// --- Branches ---------------------------------------------------------------

type SuperCabangResponse struct {
	ID         int    `json:"id"`
	Nama       string `json:"nama"`
	Alamat     string `json:"alamat"`
	NoTelp     string `json:"noTelp"`
	JamBuka    string `json:"jamBuka"`
	JamTutup   string `json:"jamTutup"`
}

type SuperCabangPerformanceResponse struct {
	IDCabang     int     `json:"idCabang"`
	NamaCabang   string  `json:"namaCabang"`
	TotalOrder   int     `json:"totalOrder"`
	TotalRevenue float64 `json:"totalRevenue"`
	TotalPaid    float64 `json:"totalPaid"`
}

type SuperCabangServiceResponse struct {
	IDLayanan        int     `json:"idLayanan"`
	NamaLayanan      string  `json:"namaLayanan"`
	SatuanLayanan    string  `json:"satuanLayanan"`
	DeskripsiLayanan *string `json:"deskripsiLayanan"`
	IDTarif          int     `json:"idTarif"`
	HargaPerSatuan   float64 `json:"hargaPerSatuan"`
	TotalItem        int     `json:"totalItem"`
	TotalRevenue     float64 `json:"totalRevenue"`
}

type CreateCabangRequest struct {
	Nama     string `json:"nama" binding:"required"`
	Alamat   string `json:"alamat" binding:"required"`
	NoTelp   string `json:"noTelp" binding:"required"`
	JamBuka  string `json:"jamBuka" binding:"required"`  // "08:00"
	JamTutup string `json:"jamTutup" binding:"required"` // "21:00"
}

type UpdateCabangRequest struct {
	Nama     *string `json:"nama"`
	Alamat   *string `json:"alamat"`
	NoTelp   *string `json:"noTelp"`
	JamBuka  *string `json:"jamBuka"`
	JamTutup *string `json:"jamTutup"`
}

type CreateTarifRequest struct {
	HargaPerSatuan   float64 `json:"hargaPerSatuan" binding:"required"`
	LayananIDLayanan int     `json:"layananIdLayanan" binding:"required"`
}

type UpdateTarifRequest struct {
	IDTarif          int      `json:"idTarif" binding:"required"`
	HargaPerSatuan   *float64 `json:"hargaPerSatuan"`
	LayananIDLayanan *int     `json:"layananIdLayanan"`
}

type SuperBranchReviewResponse struct {
	IDUlasan       int     `json:"id"`
	Rating         int     `json:"rating"`
	Komentar       string  `json:"komentar"`
	PesananID      int     `json:"pesananId"`
	PelangganNama  string  `json:"pelangganNama"`
	PegawaiNama    string  `json:"pegawaiNama"`
	LayananNama    *string `json:"layananNama"`
}

// --- Payments ---------------------------------------------------------------

type SuperPaymentResponse struct {
	ID            int       `json:"id"`
	PesananID     int       `json:"pesananId"`
	PelangganNama string    `json:"pelangganNama"`
	PelangganTelp string    `json:"pelangganNoTelp"`
	Waktu         time.Time `json:"waktu"`
	Metode        string    `json:"metode"`
	Jumlah        float64   `json:"jumlah"`
	NamaCabang    string    `json:"namaCabang"`
}

type SuperPaymentByCustomerResponse struct {
	PelangganID   int     `json:"pelangganId"`
	PelangganNama string  `json:"pelangganNama"`
	TotalOrder    int     `json:"totalOrder"`
	TotalPayment  float64 `json:"totalPayment"`
}

// --- Customers --------------------------------------------------------------

type SuperCustomerResponse struct {
	ID         int     `json:"id"`
	Nama       string  `json:"nama"`
	Email      string  `json:"email"`
	NoTelp     string  `json:"noTelp"`
	Alamat     string  `json:"alamat"`
	TotalOrder int     `json:"totalOrder"`
	TotalSpend float64 `json:"totalSpend"`
	AvgRating  float64 `json:"avgRating"`
	LastOrder  *time.Time `json:"lastOrder"`
}

type SuperCustomerOrderHistoryResponse struct {
	IDPesanan       int       `json:"id"`
	Status          string    `json:"status"`
	TotalHarga      float64   `json:"totalHarga"`
	EstimasiSelesai time.Time `json:"estimasiSelesai"`
	NamaCabang      string    `json:"namaCabang"`
}

type SuperCustomerDetailResponse struct {
	SuperCustomerResponse
	OrderHistory []SuperCustomerOrderHistoryResponse `json:"orderHistory"`
}

// --- Layanan catalog --------------------------------------------------------

type SuperLayananResponse struct {
	ID         int     `json:"id"`
	Nama       string  `json:"nama"`
	Satuan     string  `json:"satuan"`
	Deskripsi  *string `json:"deskripsi"`
}

File: frontend/src/app/lib/api/superadmin.ts
// Superadmin endpoints — every route lives under `/admin/...` and requires a
// superadmin JWT (see middleware.RequireSuperAdmin on the backend). These back
// the `(user)/admin/*` pages: dashboard, orders, services, vouchers, staffs,
// couriers, customers, branches and payments.
//
// Wire shapes mirror backend/internal/dtos/superadmin.go one-for-one. Data
// values (names, addresses, notes) stay as stored in the database; only UI
// chrome is rendered in English.

import { apiFetch } from './client';

// ============================================================================
// Types — camelCase to match the backend JSON tags.
// ============================================================================

// --- Dashboard --------------------------------------------------------------

export interface SuperStatistikUmum {
    revenueToday: number;
    revenueThisMonth: number;
    orderToday: number;
    averageRating: number;
}

export interface SuperStatistikLayanan {
    idLayanan: number;
    namaLayanan: string;
    totalPesanan: number;
    totalRevenue: number;
}

export interface SuperTopCabang {
    idCabang: number;
    namaCabang: string;
    totalOrder: number;
    totalRevenue: number;
}

export interface SuperPaymentMix {
    metode: string;
    totalEntries: number;
    persentase: number;
}

export interface SuperLiveOrdersPerCabang {
    idCabang: number;
    namaCabang: string;
    totalLiveOrders: number;
}

// --- Orders -----------------------------------------------------------------

export interface SuperPesanan {
    id: number;
    jumlahItem: number;
    status: string;
    catatan: string;
    estimasiSelesai: string;
    totalHarga: number;
    pelangganId: number;
    voucherId: number | null;
    pegawaiId: number;
    namaPelanggan: string;
    noTelpPelanggan: string;
    namaPegawai: string;
    namaCabang: string;
    idCabang: number;
}

export interface SuperOrdersStats {
    total: number;
    pickup: number;
    processing: number;
    delivery: number;
    completed: number;
}

// --- Vouchers ---------------------------------------------------------------

export interface SuperVoucher {
    id: number;
    kode: string;
    tipeDiskon: string;
    nilaiDiskon: number;
    minPembelian: number;
    berlakuHingga: string;
    kuota: number;
    terpakai: number;
}

export interface SuperVoucherStat {
    activeVoucherWeek: number;
    totalCustomerSave: number;
    waktuKadaluarsaTerdekat: string;
    waktuKadaluarsaTerdekatHours: number;
}

export interface CreateVoucherInput {
    kode: string;
    tipeDiskon: string;
    nilaiDiskon: number;
    minPembelian: number;
    berlakuHingga: string; // ISO 8601
    kuota: number;
}

export type UpdateVoucherInput = Partial<CreateVoucherInput>;

// --- Staffs -----------------------------------------------------------------

export interface SuperPegawai {
    id: number;
    nama: string;
    email: string;
    noTelp: string;
    alamat: string;
    cabangId: number;
    cabangNama: string;
}

export interface CreatePegawaiInput {
    nama: string;
    email: string;
    noTelp: string;
    alamat: string;
    cabangId: number;
}

export type UpdatePegawaiInput = Partial<CreatePegawaiInput>;

// --- Couriers ---------------------------------------------------------------

export interface SuperKurir {
    id: number;
    nama: string;
    noTelp: string;
    noPlat: string;
    tipeKendaraanId: number;
    jenisKendaraan: string;
}

export interface CreateKurirInput {
    nama: string;
    noTelp: string;
    noPlat: string;
    tipeKendaraanId: number;
}

export type UpdateKurirInput = Partial<CreateKurirInput>;

export interface SuperTipeKendaraan {
    id: number;
    jenisKendaraan: string;
}

// --- Branches ---------------------------------------------------------------

export interface SuperCabang {
    id: number;
    nama: string;
    alamat: string;
    noTelp: string;
    jamBuka: string;
    jamTutup: string;
}

export interface SuperCabangPerformance {
    idCabang: number;
    namaCabang: string;
    totalOrder: number;
    totalRevenue: number;
    totalPaid: number;
}

export interface SuperCabangService {
    idLayanan: number;
    namaLayanan: string;
    satuanLayanan: string;
    deskripsiLayanan: string | null;
    idTarif: number;
    hargaPerSatuan: number;
    totalItem: number;
    totalRevenue: number;
}

export interface CreateCabangInput {
    nama: string;
    alamat: string;
    noTelp: string;
    jamBuka: string;
    jamTutup: string;
}

export type UpdateCabangInput = Partial<CreateCabangInput>;

export interface CreateTarifInput {
    hargaPerSatuan: number;
    layananIdLayanan: number;
}

export interface UpdateTarifInput {
    idTarif: number;
    hargaPerSatuan?: number;
    layananIdLayanan?: number;
}

export interface SuperBranchReview {
    id: number;
    rating: number;
    komentar: string;
    pesananId: number;
    pelangganNama: string;
    pegawaiNama: string;
    layananNama: string | null;
}

// --- Payments ---------------------------------------------------------------

export interface SuperPayment {
    id: number;
    pesananId: number;
    pelangganNama: string;
    pelangganNoTelp: string;
    waktu: string;
    metode: string;
    jumlah: number;
    namaCabang: string;
}

// --- Customers --------------------------------------------------------------

export interface SuperCustomer {
    id: number;
    nama: string;
    email: string;
    noTelp: string;
    alamat: string;
    totalOrder: number;
    totalSpend: number;
    avgRating: number;
    lastOrder: string | null;
}

export interface SuperCustomerOrderHistory {
    id: number;
    status: string;
    totalHarga: number;
    estimasiSelesai: string;
    namaCabang: string;
}

export interface SuperCustomerDetail extends SuperCustomer {
    orderHistory: SuperCustomerOrderHistory[];
}

// --- Catalog ----------------------------------------------------------------

export interface SuperLayanan {
    id: number;
    nama: string;
    satuan: string;
    deskripsi: string | null;
}

// ============================================================================
// Fetch functions
// ============================================================================

// --- Dashboard --------------------------------------------------------------

export const getStatistikUmum = (signal?: AbortSignal) =>
    apiFetch<SuperStatistikUmum>('/admin/statistik-umum', { signal });

export const getStatistikLayanan = (signal?: AbortSignal) =>
    apiFetch<SuperStatistikLayanan[]>('/admin/statistik-layanan', { signal });

export const getTopCabang = (signal?: AbortSignal) =>
    apiFetch<SuperTopCabang[]>('/admin/statistik-cabang/top', { signal });

export const getPaymentMix = (signal?: AbortSignal) =>
    apiFetch<SuperPaymentMix[]>('/admin/statistik-pembayaran', { signal });

// --- Orders -----------------------------------------------------------------

export interface ListPesananParams {
    page?: number;
    limit?: number;
    status?: string;
    search?: string;
}

export const listPesanan = (params: ListPesananParams = {}, signal?: AbortSignal) =>
    apiFetch<SuperPesanan[]>(`/admin/statistik-pesanan${buildQuery(params)}`, { signal });

export const getOrdersStats = (signal?: AbortSignal) =>
    apiFetch<SuperOrdersStats>('/admin/orders/stats', { signal });

// --- Services ---------------------------------------------------------------

export const listServiceStats = (signal?: AbortSignal) =>
    apiFetch<SuperStatistikLayanan[]>('/admin/services', { signal });

// --- Vouchers ---------------------------------------------------------------

export const listVouchers = (signal?: AbortSignal) =>
    apiFetch<SuperVoucher[]>('/admin/vouchers', { signal });

export const getVouchersStatistik = (signal?: AbortSignal) =>
    apiFetch<SuperVoucherStat>('/admin/vouchers-statistik', { signal });

export const createVoucher = (input: CreateVoucherInput) =>
    apiFetch<SuperVoucher>('/admin/vouchers', { method: 'POST', body: input });

export const getVoucher = (id: number, signal?: AbortSignal) =>
    apiFetch<SuperVoucher>(`/admin/vouchers/${id}`, { signal });

export const updateVoucher = (id: number, input: UpdateVoucherInput) =>
    apiFetch<SuperVoucher>(`/admin/vouchers/${id}`, { method: 'PUT', body: input });

export const deleteVoucher = (id: number) =>
    apiFetch<{ deleted: boolean }>(`/admin/vouchers/${id}`, { method: 'DELETE' });

// --- Staffs -----------------------------------------------------------------

export const listPegawai = (signal?: AbortSignal) =>
    apiFetch<SuperPegawai[]>('/admin/staffs', { signal });

export const createPegawai = (input: CreatePegawaiInput) =>
    apiFetch<SuperPegawai>('/admin/staffs', { method: 'POST', body: input });

export const updatePegawai = (id: number, input: UpdatePegawaiInput) =>
    apiFetch<SuperPegawai>(`/admin/staffs/${id}`, { method: 'PUT', body: input });

export const deletePegawai = (id: number) =>
    apiFetch<{ deleted: boolean }>(`/admin/staffs/${id}`, { method: 'DELETE' });

// --- Couriers ---------------------------------------------------------------

export const listKurir = (signal?: AbortSignal) =>
    apiFetch<SuperKurir[]>('/admin/couriers', { signal });

export const createKurir = (input: CreateKurirInput) =>
    apiFetch<SuperKurir>('/admin/couriers', { method: 'POST', body: input });

export const updateKurir = (id: number, input: UpdateKurirInput) =>
    apiFetch<SuperKurir>(`/admin/couriers/${id}`, { method: 'PUT', body: input });

export const deleteKurir = (id: number) =>
    apiFetch<{ deleted: boolean }>(`/admin/couriers/${id}`, { method: 'DELETE' });

export const listTipeKendaraan = (signal?: AbortSignal) =>
    apiFetch<SuperTipeKendaraan[]>('/admin/tipe-kendaraan', { signal });

// --- Branches ---------------------------------------------------------------

export const listCabang = (signal?: AbortSignal) =>
    apiFetch<SuperCabang[]>('/admin/branch', { signal });

export const getCabang = (id: number, signal?: AbortSignal) =>
    apiFetch<SuperCabang>(`/admin/branch/${id}`, { signal });

export const getBranchPerformance = (signal?: AbortSignal) =>
    apiFetch<SuperCabangPerformance[]>('/admin/branch/detail/performance', { signal });

export const getBranchServices = (id: number, signal?: AbortSignal) =>
    apiFetch<SuperCabangService[]>(`/admin/branch/${id}/services`, { signal });

export const getBranchReviews = (
    id: number,
    params: { page?: number; limit?: number } = {},
    signal?: AbortSignal,
) => apiFetch<SuperBranchReview[]>(`/admin/branch/${id}/reviews${buildQuery(params)}`, { signal });

export const createCabang = (input: CreateCabangInput) =>
    apiFetch<SuperCabang>('/admin/branch', { method: 'POST', body: input });

export const updateCabang = (id: number, input: UpdateCabangInput) =>
    apiFetch<SuperCabang>(`/admin/branch/${id}`, { method: 'PUT', body: input });

export const deleteCabang = (id: number) =>
    apiFetch<{ deleted: boolean }>(`/admin/branch/${id}`, { method: 'DELETE' });

export const createTarif = (cabangId: number, input: CreateTarifInput) =>
    apiFetch<{ created: boolean }>(`/admin/branch/${cabangId}/services`, {
        method: 'POST',
        body: input,
    });

export const updateTarif = (cabangId: number, input: UpdateTarifInput) =>
    apiFetch<{ updated: boolean }>(`/admin/branch/${cabangId}/services`, {
        method: 'PUT',
        body: input,
    });

// --- Payments ---------------------------------------------------------------

export interface ListPaymentsParams {
    method?: string;
    search?: string;
    page?: number;
    limit?: number;
}

export const listPayments = (params: ListPaymentsParams = {}, signal?: AbortSignal) =>
    apiFetch<SuperPayment[]>(`/admin/payments${buildQuery(params)}`, { signal });

// --- Customers --------------------------------------------------------------

export interface ListCustomersParams {
    search?: string;
    page?: number;
    limit?: number;
}

export const listCustomers = (params: ListCustomersParams = {}, signal?: AbortSignal) =>
    apiFetch<SuperCustomer[]>(`/admin/customers${buildQuery(params)}`, { signal });

export const getCustomer = (id: number, signal?: AbortSignal) =>
    apiFetch<SuperCustomerDetail>(`/admin/customers/${id}`, { signal });

// --- Catalog ----------------------------------------------------------------

export const listLayanan = (signal?: AbortSignal) =>
    apiFetch<SuperLayanan[]>('/admin/layanan', { signal });

// ============================================================================
// Helpers
// ============================================================================

function buildQuery(params: object): string {
    const search = new URLSearchParams();
    for (const [key, value] of Object.entries(params as Record<string, string | number | undefined>)) {
        if (value !== undefined && value !== '') search.set(key, String(value));
    }
    const qs = search.toString();
    return qs ? `?${qs}` : '';
}
